### ISSUE TITLE
A proper fix for v1.0.2 on real hardware & PS3 | xdelta patch addendum

### ISSUE THREAD

**Issue by swosho** [wagrenier/Zero2Undub#13]
Here's a temporary fix for the freezes on real hardware (also mentioned in https://github.com/wagrenier/Zero2Undub/issues/12), as the currently available "Safe Undub" option for the real hardware compatibility results in a partial undub.

Changes:
1. Corrected the total ISO size in the ISO ToC
2. Corrected the IMG_BD.BIN size in the ISO ToC 
3. Placed the dummy files (shrunken down to 2 kb) after the IMG_BD.BIN like in the original image so that IMG_BD.BIN doesn't overlap with them; ofc also reflected the dummy files' new offsets and sizes in the ISO ToC
4. Appended back the original image's end padding and the EoF sector to the end of the image

This turns the ISO "legit," and just that seems to resolve the freezes that were occurring on PS2 & PS3.

[Fatal Frame II Undub v1.0.2 fix.zip](https://github.com/user-attachments/files/17280953/Fatal.Frame.II.Undub.v1.0.2.fix.zip)
Apply to the v1.0.2 undubbed ISO (CRC32: `67F461BD` / MD5: `8B33F35838CE97E73D311E14B1A3385B`)

Late edit: Unrelated, but the MIA European undub patch [can be found here](https://archive.org/details/project_zero_2_ps2_undub_patch).

**Comment by wagrenier** [comment]
I believe as of now my patch already does steps 3) and 4). My undub most likely crashes because I do not correctly adjust the file sizes for the ISO. @swosho 

I'll look at the offsets of an ISO to achieve that and push that into a branch. Would you be able to test the patch?

**Comment by swosho** [comment]
> I believe as of now my patch already does steps 3) and 4)

If by that you mean v1.0.2, then that's not quite the case. IMG_BD.BIN's appended data at 0xC23CB000 just overwrites everything starting from there, which is: the dummy files, the end padding / EoF postgap, and the EoF sector. I guess the stuff that the patcher currently overwrites was meant to be shifted forward, then?

> Would you be able to test the patch?

The undub with the fix applied is confirmed to run issue-free on an actual PS2, I think there's no need for further testing if you can simply compare the upcoming patch version against this temporary addendum.
Here's the same fix as in the OP but with the dummy files' sizes kept intact so it would be identical to an ISO with the ToC sorted out and all the end data shifted forward instead being of overwritten: [Fatal Frame II Undub v1.0.2 fix-norip.zip](https://github.com/user-attachments/files/17304326/Fatal.Frame.II.Undub.v1.0.2.fix-norip.zip)

**Comment by Bag-Head** [comment]
If it helps, I'm testing this on real hardware right now, it's already looking good as the first freeze I would have on real hardware (the first cutscene after entering the first building) no longer happens, I'm gonna play a bit more but I'm likely to use the fixed patch to pick up my savefile in Chapter 5!

Great work fixing this

**Comment by wagrenier** [comment]
Thanks to both for the information with that I'll work on the fix and it should be out today or tomorrow for FF2 and maybe next week for FF1

**Comment by wagrenier** [comment]
I am done writing the code needed for this fix, I'll be testing it tomorrow.

**Comment by wagrenier** [comment]
I need to make further adjustments so next week I'll have the new version released. @swosho that tool did you use to change the file sizes? My current patches does not read files so the process of manually patch the iso is very tedious. I think my best solution would be rewriting the patcher

**Comment by swosho** [comment]
I did those edits plainly in a hex editor, since there were so few values to correct after all ¯\_(ツ)_/¯
You'd normally parse and fix the ISO9660 header, and then do the same for the UDF descriptors following it... but in this case, I think it's not too much of a "hack" to just manually hardcode the offsets for the needed edits and write predetermined values there.

**Comment by wagrenier** [comment]
Yeah I'll try to find an ISO9660 parser for C++ or C#, would most likely mean a rewrite tho, which will take a bit more time

**Comment by DarthMotzkus** [comment]
@swosho @wagrenier you guys ROCKS!

**Comment by DarthMotzkus** [comment]
@wagrenier forgot to merge on main?

**Comment by wagrenier** [comment]
@DarthMotzkus No, I ended up adding some stuff but in order to fully fix it it got way outta hand because of how I get the offsets for each file I need and to rewrite the iso correctly. Instead, I started looking for ISO reader/writer for C++ but could not find any. I have reached a point where I'd need to rewrite the project and I don't have much time for that right now, as most of my free time goes into the [decompilation project of FF1](https://github.com/Mikompilation/Himuro), which I value a lot more than the undub. I will eventually fix it tho...
